google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQueryResultSetImpl.java

```diff
@@ -0,0 +1,554 @@
+
+    private boolean isEndOfStream(T cursor) {
+      return cursor instanceof ConnectionImpl.EndOfFieldValueList;
+    }
+
+    @Override
+    public Object getObject(String fieldName) throws SQLException {
+      if (fieldName == null) {
+        throw new SQLException("fieldName can't be null");
+      }
+      if (cursor == null) {
+        return null;
+      } else if (cursor instanceof FieldValueList) {
+        FieldValue fieldValue = ((FieldValueList) cursor).get(fieldName);
+        return fieldValue.getValue() == null ? null : fieldValue.getValue();
+      } else { // Data received from Read API (Arrow)
+        Tuple<Map<String, Object>, Boolean> curTuple = (Tuple<Map<String, Object>, Boolean>) cursor;
+        if (!curTuple.x().containsKey(fieldName)) {
+          throw new SQLException(String.format("Field %s not found", fieldName));
+        }
+        return curTuple.x().get(fieldName);
+      }
+    }
+
+    @Override
+    public Object getObject(int columnIndex) throws SQLException {
+      if (cursor == null) {
+        return null;
+      } else if (cursor instanceof FieldValueList) {
+        FieldValue fieldValue = ((FieldValueList) cursor).get(columnIndex);
+        return fieldValue.getValue() == null ? null : fieldValue.getValue();
+      } else { // Data received from Read API (Arrow)
+        return getObject(schemaFieldList.get(columnIndex).getName());
+      }
+    }
+
+    @Override
+    public String getString(String fieldName) throws SQLException {
+      if (fieldName == null) {
+        throw new SQLException("fieldName can't be null");
+      }
+      if (cursor == null) {
+        return null;
+      } else if (cursor instanceof FieldValueList) {
+        FieldValue fieldValue = ((FieldValueList) cursor).get(fieldName);
+        if (fieldValue.getValue() == null) {
+          return null;
+        } else if (fieldValue
+            .getAttribute()
+            .equals(FieldValue.Attribute.REPEATED)) { // Case for Arrays
+          return fieldValue.getValue().toString();
+        } else {
+          return fieldValue.getStringValue();
+        }
+      } else { // Data received from Read API (Arrow)
+        Tuple<Map<String, Object>, Boolean> curTuple = (Tuple<Map<String, Object>, Boolean>) cursor;
+        if (!curTuple.x().containsKey(fieldName)) {
+          throw new SQLException(String.format("Field %s not found", fieldName));
+        }
+        Object currentVal = curTuple.x().get(fieldName);
+        if (currentVal == null) {
+          return null;
+        } else if (currentVal instanceof JsonStringArrayList) { // arrays
+          JsonStringArrayList jsnAry = (JsonStringArrayList) currentVal;
+          return jsnAry.toString();
+        } else if (currentVal instanceof LocalDateTime) {
+          LocalDateTime dateTime = (LocalDateTime) currentVal;
+          return dateTime.toString();
+        } else {
+          Text textVal = (Text) currentVal;
+          return textVal.toString();
+        }
```

Comment:
Since nothing custom is done for those three types, it's safe to check if `currentVal` is one of either of those three, and then just call `toString()` -- the overloading will take care of the specific type implementation.

Reply:
Can I request some more insight here? I think calling `toString` on `currentVal` would call `Object`'s `toString` , hence explicit typecasting will be required.

Reply:
I actually used the wrong term in the original comment, but the point stands -- it should just be Java's method *overriding* (same signature), not "overloading" (different signature). When class methods have the same signature as methods inherited from a superclass, Java's implementation of polymorphism will mean it will invoke the method on the actual object type, not on the reference handle.

So, if a `toString()` implementation is available on `Text`, `LocalDateTime` and `JsonStringArrayList` classes, the object's specific implementation will be used. The only time `Object.toString()` implementation will be called is if none of the superclasses of the specific object provide a `toString()` implementation.